mlflow/utils/model_utils.py

```diff
@@ -191,3 +194,78 @@ def _validate_onnx_session_options(onnx_session_options):
                     f"Value for key {key} in onnx_session_options should be >= 0, not {value}"
                 )
+
+
+def _get_overridden_pyfunc_model_config(
+    pyfunc_config: Dict[str, Any], load_config: Dict[str, Any], logger
+) -> Dict[str, Any]:
+    """
+    Updates the inference configuration according to the model's configuration and the overrides.
+    Only arguments already present in the inference configuration can be updated. The environment
+    variable ``MLFLOW_PYFUNC_INFERENCE_CONFIG`` can also be used to provide additional inference
+    configuration.
+    """
+
+    overrides = {}
+    if env_overrides := os.getenv("MLFLOW_PYFUNC_INFERENCE_CONFIG"):
+        logger.debug(
+            "Inference configuration is being loaded from ``MLFLOW_PYFUNC_INFERENCE_CONFIG``"
+            " environ."
+        )
+        overrides.update(dict(json.loads(env_overrides)))
+
+    if load_config:
+        overrides.update(load_config)
+
+    if not overrides:
+        return pyfunc_config
+
+    if not pyfunc_config:
+        logger.warning(
+            f"Argument(s) {', '.join(overrides.keys())} were ignored since the model's ``pyfunc``"
+            " flavor doesn't accept model configuration. Use ``model_config`` when logging"
+            " the model to allow it."
+        )
+
+        return None
+
+    allowed_config = {key: value for key, value in overrides.items() if key in pyfunc_config.keys()}
+
+    if len(allowed_config) < len(overrides):
+        ignored_keys = set(overrides.keys()) - set(allowed_config.keys())
+        logger.warning(
+            f"Argument(s) {', '.join(ignored_keys)} were ignored since they are not valid keys in"
+            " the corresponding section of the ``pyfunc`` flavor. Use ``model_config`` when"
+            " logging the model to include the keys you plan to indicate. Current allowed"
+            " configuration includes"
+            f" {', '.join(pyfunc_config.keys())}"
+        )
+
+    pyfunc_config.update(allowed_config)
+    return pyfunc_config
+
+
+def _validate_pyfunc_model_config(model_config):
+    """
+    Validates the values passes in the model_config section. There are no typing
+    restrictions but we require them being JSON-serializable.
+    """
+
+    def is_jsonable(value):
+        try:
+            json.dumps(value)
+            return True
+        except (TypeError, OverflowError):
+            return False
+
+    if model_config:
+        if isinstance(model_config, dict) and all(
+            isinstance(key, str) for key in model_config.keys()
+        ):
+            if not is_jsonable(model_config):
+                raise MlflowException(
+                    "Some of the values indicated in ``model_config`` are not "
+                    "supported. Only JSON-serializable data types can be indicated."
+                )
+        else:
+            raise MlflowException("``model_config`` has to be of type ``dict`` with string keys.")
```

Comment:
Could we short-circuit the return and handle the two separate errors that might arise:
1. A dict with non-string keys was passed (invalid model_config structure) 
2. A dict with string keys and non-encodable objects was passed

```suggestion
def _validate_pyfunc_model_config(model_config):
    """
    Validates the values passes in the model_config section. There are no typing
    restrictions but we require them being JSON-serializable.
    """

    if not model_config:
        return

    if not isinstance(model_config, dict) or not all(isinstance(key, str) for key in model_config):
        raise MlflowException("An invalid ``model_config`` structure was passed. ``model_config`` must be of type ``dict`` with string keys.")
    
    try:
        json.dumps(model_config)
    except(TypeError, OverflowError):
        raise MlflowException("Values in the provided ``model_config`` are of an unsupported type. Only JSON-serializable data types can be provided as values.")
```

Or something similar to that?